spec/compiler/type_inference/instance_var_spec.cr

```diff
@@ -2157,6 +2157,154 @@ describe "Type inference: instance var" do
       )) { int32 }
   end
 
+  it "infers from class method that has type annotation" do
+    assert_type(%(
+      class Bar
+        def self.bar : Bar
+          Bar.new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from class method that has type annotation, in generic class" do
+    assert_type(%(
+      class Bar
+        def self.bar : Bar
+          Bar.new
+        end
+      end
+
+      class Foo(T)
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo(Int32).new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from generic class method that has type annotation" do
+    assert_type(%(
+      class Bar(T)
+        def self.bar : self
+          Bar(T).new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar(Int32).bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { (types["Bar"] as GenericClassType).instantiate([int32] of TypeVar) }
+  end
+
+  it "infers from generic class method that has type annotation, without instantiating" do
+    assert_type(%(
+      class Bar(T)
+        def self.bar : Int32
+          1
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { int32 }
+  end
+
+  it "infers from class method that has type annotation, with overload" do
+    assert_type(%(
+      class Baz
+      end
+
+      class Bar
+        def self.bar : Baz
+          Baz.new
+        end
+
+        def self.bar(x) : Bar
+          Bar.new
+        end
+
+        def self.bar(x) : Baz
+          yield
+          Bar.new
+        end
+      end
+
+      class Foo
+        def initialize
+          @bar = Bar.bar(1)
+        end
+
+        def bar
+          @bar
+        end
+      end
+
+      Foo.new.bar
+      )) { types["Bar"] }
+  end
+
+  it "infers from class method that has type annotation, with multiple overloads matching, all with the same type" do
```

Comment:
Code:

```crystal
class Bar
  def self.bar : Int32
    "hello"
  end
end

class Foo
  getter x

  def initialize
    @x = Bar.bar
  end
end

Foo.new.x
```

Error:

```
Error in ./bar.cr:15: instantiating 'Foo:Class#new()'

Foo.new.x
    ^~~

instantiating 'Foo#initialize()'

in ./bar.cr:11: instantiating 'Bar:Class#bar()'

    @x = Bar.bar
             ^~~

in ./bar.cr:2: type must be Int32, not String

  def self.bar : Int32
           ^~~
```

Methods are looked up just by arguments count and whether they have a block or not. If later something is not as expected a union isn't created, an error saying you are assigning an unexpected type happens.